notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/NetworkStatusReceiver.java

```diff
@@ -0,0 +1,40 @@
+package com.microsoft.windowsazure.messaging.notificationhubs;
+
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.net.ConnectivityManager;
+import android.net.NetworkInfo;
+
+/**
+ * Synchronizes updates to the tags or other installation aware items on the device
+ * when the device comes back online.
+ */
+public class NetworkStatusReceiver extends BroadcastReceiver {
+
+    private final NotificationHub mNotificationHub;
+
+    public NetworkStatusReceiver() {
+        mNotificationHub = NotificationHub.getInstance();
+    }
+
+    public NetworkStatusReceiver(NotificationHub nh) {
+        mNotificationHub = nh;
+    }
+
+    @Override
+    public void onReceive(Context context, Intent intent) {
+        ConnectivityManager cm =
+                (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
+        NetworkInfo activeNetwork = cm.getActiveNetworkInfo();
+        boolean isConnected = activeNetwork != null &&
+                activeNetwork.isConnectedOrConnecting();
+        if (isConnected) {
+            try {
+                mNotificationHub.reinstallInstance();
```

Comment:
Does this only get called when the status changes? I almost think we should just pass the connected state to the NotificationHub class and let it determine if the state has changed and/or requires a "reinstall".

Reply:
Because each component of the what will build up the Installation is independent, it's a lot easier to just build up a new Installation each time anything's happened and see if the end product is the same as the last thing to hit the wire. Saves each `Visitor` from needing to implement an `isStale` method or similar.

Reply:
I'm not suggesting changing the visitor design, but we should understand if this callback is going to fire when we expect.

Also, right now we attempt to "debounce" the change events in the NotificationAdapter, but if someone replaces that with their own implementation we should still make an effort to only call `updateInstallation` when we actually have an update, otherwise we just push the problem to the app developer.

Reply:
Totally fair. The way we were going to handle this was to continue to expose the `DebounceInstallationAdapter` publicly. Folks can decorate their `InstallationAdapter`s with a single call. IMHO, this is a good balance of giving folks additional control of when things are going to hit the wire, but without burdening them with rewriting basic functionality.

For preview1, I've hidden `InstallationAdapter` and everything related to it, to give us more time to think through this and #55.